contracts/ERC20MultiDelegate.sol

```diff
@@ -0,0 +1,251 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.2;
+
+import {Address} from "@openzeppelin/contracts/utils/Address.sol";
+
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts/token/ERC1155/ERC1155.sol";
+import "@openzeppelin/contracts/token/ERC20/extensions/ERC20Votes.sol";
+
+/**
+ * @dev A child contract which will be deployed by the ERC20MultiDelegate utility contract
+ * This is a proxy delegator contract to vote given delegatee on behalf of original delegator
+ */
+contract ERC20ProxyDelegator {
+    constructor(ERC20Votes _token, address _delegatee) {
+        _token.approve(msg.sender, type(uint256).max);
+        _token.delegate(_delegatee);
+    }
+}
+
+/**
+ * @dev A utility contract to let delegators to pick multiple delegatee
+ */
+contract ERC20MultiDelegate is ERC1155, Ownable {
+    using Address for address;
+
+    ERC20Votes public token;
+
+    struct DelegateeAmount {
+        address delegatee;
+        uint256 amount;
+    }
+
+    struct SourceTargetDelegatee {
+        address source;
+        address target;
+    }
+
+    /**
+     * @dev Constructor.
+     * @param _token The ERC20 token address
+     * @param _metadata_uri ERC1155 metadata uri
+     */
+    constructor(
+        ERC20Votes _token,
+        string memory _metadata_uri
+    ) ERC1155(_metadata_uri) {
+        token = _token;
+    }
+
+    /**
+     * @dev Deposits and delegates voting power to the multiple delegatees.
+     * @param delegateeAmounts The list of delegatee addresses and corresponding list of ERC20 voting power amount amounts to delegate.
+     */
+    function depositMulti(
+        DelegateeAmount[] calldata delegateeAmounts
+    ) external {
+        uint256 delegateesLength = delegateeAmounts.length;
+
+        require(
+            delegateesLength > 0,
+            "DepositMulti: You should pick at least one delegatee"
+        );
+
+        uint256[] memory ids = new uint256[](delegateesLength);
+        uint256[] memory amounts = new uint256[](delegateesLength);
+
+        for (uint256 index = 0; index < delegateesLength; index++) {
+            address delegatee = delegateeAmounts[index].delegatee;
+            uint256 amount = delegateeAmounts[index].amount;
+
+            createProxyDelegatorAndTransfer(delegatee, amount);
+
+            ids[index] = uint256(uint160(delegatee));
+            amounts[index] = amount;
+        }
+
+        mintBatch(msg.sender, ids, amounts);
+    }
+
+    /**
+    * @dev Re-delegates voting power between delegatees.
+    * @param delegateePairs The list of current delegatee addresses (from which the voting power is withdrawn) 
+    * and target delegatee addresses (which the voting power will be delegated).
+    */
+    function reDeposit(SourceTargetDelegatee[] calldata delegateePairs) external {
```

Comment:
One thing I am unsure about this method, should we map the transfer as 1:1, so for each `source` we will require a `target` address. or should it give freedom to user so the delegation can be many to many, like; for 2 `source` address, user can define let's say 3 `target` address with arbitrary `amount` to split or move the voting power?

Reply:
okay I change it as I described above, but instead of three different arrays, I required user to provide two array of structs;
```
[[source_address, withdraw_amount] ...], [[target_address, target_amount]...]
```
both array sizes can be arbitrary so user can redeposit from many to many. 

In case user choose to withdraw more amount from source addresses in total but deposit less to targets in total, the remaining amount will be reimbursed to the caller.